File: nuido_flow_shopify/flows/shopify/shopify_auth_node.py
# THIS FILE IS A PART OF PUBLIC REPOSITORY https://github.com/yonitjio/exploring-odoo
#
# This software is released under the MIT License.
# https://opensource.org/licenses/MIT
#
# THIS SOFTWARE IS EXPERIMENTAL AND FOR EDUCATIONAL PURPOSE ONLY.
# DO NOT USE IT IN PRODUCTION.
import hmac
import hashlib
import base64

from werkzeug.exceptions import Unauthorized
from odoo.http import request
from odoo.addons.nuido_flow.flows.core.base_node import BaseNode

class ShopifyAuthNode(BaseNode):
    def _calc_shopify_secret(self, secret):
        data = request.httprequest.get_data(cache=False)
        hmac_digest = hmac.new(secret, data, hashlib.sha256).digest()
        shopify_secret = base64.b64encode(hmac_digest).decode()

        return shopify_secret

    def _process(self, params) -> any:
        super()._process(params)

        secret = str(self.definition["secret"]).encode()
        shopify_secret = self._calc_shopify_secret(secret)

        if shopify_secret != request.httprequest.headers.environ["HTTP_X_SHOPIFY_HMAC_SHA256"]:
            self.next_node_info = None
            if self.definition["raise_error"]:
                raise Unauthorized()

        return params

File: nuido_flow_shopify/static/src/components/shopify/shopify_auth_node.js
// THIS FILE IS A PART OF PUBLIC REPOSITORY https://github.com/yonitjio/exploring-odoo
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
//
// THIS SOFTWARE IS EXPERIMENTAL AND FOR EDUCATIONAL PURPOSE ONLY.
// DO NOT USE IT IN PRODUCTION.
import { Node } from "@nuido/components/node";
import { TextInputDialogInput } from "@nuido_base/components/dialogs/text_input_dialog_input";
export class ShopifyAuthNode extends Node {
    onSecretChanged(value) {
        this.props.node.secret = value;
    }
    onRaiseErrorChanged() {
        if (this.props.node.raise_error) {
            this.props.node.raise_error = false;
        }
        else {
            this.props.node.raise_error = true;
        }
    }
    get raiseErrorInputId() {
        return `label-${this.props.node.id}-raise-error`;
    }
}
ShopifyAuthNode.template = "nuido_flow_shopify.shopify-auth-node";
ShopifyAuthNode.components = {
    ...Node.components,
    TextInputDialogInput
};
